pkg/pmem-csi-driver/nodeserver.go

```diff
@@ -88,57 +101,92 @@ func (ns *nodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublis
-	if err != nil && !os.IsNotExist(err) {
-		return nil, status.Error(codes.Internal, "validate target path: "+err.Error())
-	}
-	if !notMnt {
-		// TODO(https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/issues/95): check if mount is compatible. Return OK if it is, or appropriate error.
-		/*
-			1) Target Path MUST be the vol referenced by vol ID
-			2) VolumeCapability MUST match
-			3) Readonly MUST match
-		*/
-		return &csi.NodePublishVolumeResponse{}, nil
-	}
+	srcPath := req.GetStagingTargetPath()
+	ephemeral := false
 
-	if err := os.Mkdir(targetPath, os.FileMode(0755)); err != nil {
-		// Kubernetes is violating the CSI spec and creates the
-		// directory for us
-		// (https://github.com/kubernetes/kubernetes/issues/75535). We
-		// allow that by ignoring the "already exists" error.
-		if !os.IsExist(err) {
-			return nil, status.Error(codes.Internal, "make target dir: "+err.Error())
+	// Kubernetes v1.16+ would request ephemeral volumes via VolumeContext
+	val, ok := req.GetVolumeContext()[volumeContextEphemeral]
+	if ok {
+		b, err := strconv.ParseBool(val)
+		if err != nil {
+			return nil, status.Error(codes.InvalidArgument, err.Error())
 		}
+		ephemeral = b
+	} else {
+		// If met all below conditions we still treat as a ephemeral volume request:
+		// 1) No Volume found with given volume id
+		// 2) No provisioner info found in VolumeContext "storage.kubernetes.io/csiProvisionerIdentity"
+		// 3) No StagingPath in the request
+		device, _ := ns.cs.dm.GetDevice(req.VolumeId)
+		_, ok := req.GetVolumeContext()[volumeProvisionerIdentity]
+		ephemeral = device == nil && !ok && len(srcPath) == 0
 	}
 
-	readOnly := req.GetReadonly()
-	attrib := req.GetVolumeContext()
-	mountFlags := req.GetVolumeCapability().GetMount().GetMountFlags()
+	var mountOptions []string
 
-	klog.V(3).Infof("NodePublishVolume: id:%s targetpath:%v stagingtargetpath:%v readonly:%v attributes:%v mountflags:%v",
-		req.GetVolumeId(), targetPath, stagingtargetPath, readOnly, attrib, mountFlags)
+	if ephemeral {
+		val, ok := req.GetVolumeContext()["size"]
+		if !ok {
+			return nil, status.Error(codes.InvalidArgument, "size not found in volume context")
+		}
+
+		size, err := resource.ParseQuantity(val)
+		if err != nil {
+			return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("Failed to parse given size '%s': %s", val, err.Error()))
+		}
+
+		// Create new device
+		params := req.GetVolumeContext()
+		params[pmemParameterKeyPersistencyModel] = string(pmemPersistencyModelEphemeral)
+		createReq := &csi.CreateVolumeRequest{
+			VolumeCapabilities: []*csi.VolumeCapability{req.VolumeCapability},
+			Name:               req.VolumeId,
+			CapacityRange:      &csi.CapacityRange{RequiredBytes: size.Value()},
+			Parameters:         params,
+		}
 
-	options := []string{"bind"}
-	if readOnly {
-		options = append(options, "ro")
+		createResp, err := ns.cs.CreateVolume(ctx, createReq)
```

Comment:
We need full validation of the volume parameters. Unknown (or miss-spelled) parameters must trigger an error. Otherwise user errors may silently lead to a different than intended outcome.

As a testcase, I tried with `foo: bar` added to `pmem-app-ephemeral.yaml` and that was silently ignored.

Reply:
And we also need an automated test case for this.

Reply:
Beware that we enable pod info on mount and thus have to ignore *all* `csi.storage.k8s.io/*` entries. Unknown fields may get added under that namespace in the future.

Reply:
Done, Now we validate the volume context and raise an error if occurred any unexpected keys. Testcase is pending.